src/dsa/rpo_falcon512/mod.rs

```diff
@@ -0,0 +1,60 @@
+use crate::{
+    hash::rpo::Rpo256,
+    utils::{
+        collections::Vec, ByteReader, ByteWriter, Deserializable, DeserializationError,
+        Serializable,
+    },
+    Felt, StarkField, Word, ZERO,
+};
+
+#[cfg(feature = "std")]
+mod ffi;
+
+mod error;
+mod keys;
+mod polynomial;
+mod signature;
+
+pub use error::FalconError;
+pub use keys::{KeyPair, PublicKey};
+pub use polynomial::Polynomial;
+pub use signature::Signature;
+
+// CONSTANTS
+// ================================================================================================
+
+// The Falcon modulus.
+const MODULUS: u16 = 12289;
+const MODULUS_MINUS_1_OVER_TWO: u16 = 6144;
+
+// The Falcon parameters for Falcon-512. This is the degree of the polynomial `phi := x^n + 1`
+// defining the ring Z_q[x]/(phi).
+const N: usize = 512;
+const LOG_N: usize = 9;
+
+/// Length of nonce used for key-pair generation.
+const NONCE_LEN: usize = 40;
+
+/// Number of filed elements used to encode a nonce.
+const NONCE_ELEMENTS: usize = 8;
+
+/// Public key length as a u8 vector.
+const PK_LEN: usize = 897;
+
+/// Secret key length as a u8 vector.
+const SK_LEN: usize = 1281;
+
+/// Signature length as a u8 vector.
+const SIG_LEN: usize = 625;
```

Comment:
A couple of questions:
- In the C interface docs, we specify that the signature can be up to 666 bytes long. However, here the combination of signature + nonce is only 665 bytes. Why can we discard the last byte?
- It seems like we ignore the first byte of the returned signature (nonce starts at `nonce[1..41]`. What is this byte and why can we ignore it?

Reply:
The reference implementation (i.e., the one PQClean is based on) used the first byte to specify the security level (i.e. N = 512 vs N = 1024 vs N = ...) and it stores the `log(N)`. In the PQClean version, this becomes unnecessary as there are only two implementations provided and throughout the code the first byte is used just as a sanity check. For us, this becomes even more redundant as we have only one implementation and, even if we had more than one implementation, the Rust code provides enough guarantees that things will work fine.
This is why in an earlier commit the signature was defined as  `Signature::new(&sig[1..41], &sig[41..], self.public_key)` and, based on this, the signature + nonce is 665 versus the 666 in the PQClean side. In the latest commit, we now have `Ok(Signature { sig, pk: self.public_key })` and I am not sure this takes the above into account.  
I propose that we use we use `(665 + 1)` to keep things uniform on both sides.

Reply:
> I propose that we use we use `(665 + 1)` to keep things uniform on both sides.

Agreed!